api/pkg/apis/v1alpha1/managers/solution/solution-manager.go

```diff
@@ -420,132 +421,154 @@ func (s *SolutionManager) Reconcile(ctx context.Context, deployment model.Deploy
-				summary.UpdateTargetResult(step.Target, model.TargetResultSpec{Status: targetResultStatus, Message: targetResultMessage, ComponentResults: componentResults}) // TODO: this keeps only the last error on the target
-				time.Sleep(5 * time.Second)                                                                                                                                   //TODO: make this configurable?
+				provider = override
 			}
-		}
-		if stepError != nil {
-			log.ErrorfCtx(ctx, " M (Solution): failed to execute deployment step: %+v", stepError)
 
-			successCount := 0
-			for _, v := range targetResult {
-				successCount += v
+			if previousDesiredState != nil {
+				testState := MergeDeploymentStates(&previousDesiredState.State, currentState)
+				if s.canSkipStep(ctx, step, step.Target, provider.(tgt.ITargetProvider), previousDesiredState.State.Components, testState) {
+					log.InfofCtx(ctx, " M (Solution): skipping step with role %s on target %s", step.Role, step.Target)
+					targetResult[step.Target] = 1
+					planSuccessCount++
+					continue
+				}
 			}
-			summary.CurrentDeployed += successCount
-			summary.SuccessCount = successCount
-			summary.AllAssignedDeployed = plannedCount == planSuccessCount
-			err = stepError
-			return summary, err
-		}
-		planSuccessCount++
-		summary.CurrentDeployed += len(step.Components)
-		err = s.saveSummaryProgress(ctx, deployment.Instance.ObjectMeta.Name, deployment.Generation, deployment.Hash, summary, namespace)
-		if err != nil {
-			log.ErrorfCtx(ctx, " M (Solution): failed to save summary progress: %+v", err)
-			return summary, err
+			log.DebugfCtx(ctx, " M (Solution): applying step with Role %s on target %s", step.Role, step.Target)
+			someStepsRan = true
+			retryCount := 1
+			//TODO: set to 1 for now. Although retrying can help to handle transient errors, in more cases
+			// an error condition can't be resolved quickly.
+			var stepError error
+			var componentResults map[string]model.ComponentResultSpec
+
+			// for _, component := range step.Components {
+			// 	for k, v := range component.Component.Properties {
+			// 		if strV, ok := v.(string); ok {
+			// 			parser := api_utils.NewParser(strV)
+			// 			eCtx := s.VendorContext.EvaluationContext.Clone()
+			// 			eCtx.DeploymentSpec = deployment
+			// 			eCtx.Component = component.Component.Name
+			// 			val, err := parser.Eval(*eCtx)
+			// 			if err == nil {
+			// 				component.Component.Properties[k] = val
+			// 			} else {
+			// 				log.ErrorfCtx(ctx, " M (Solution): failed to evaluate property: %+v", err)
+			// 				summary.SummaryMessage = fmt.Sprintf("failed to evaluate property '%s' on component '%s: %s", k, component.Component.Name, err.Error())
+			// 				s.saveSummary(ctx, deployment, summary)
+			// 				return summary, err
+			// 			}
+			// 		}
+			// 	}
+			// }
+
+			for i := 0; i < retryCount; i++ {
+				select {
+				case <-ctx.Done():
```

Comment:
I'm not sure if we should allow cancelation at per-component level, as it may leave the target at undesired state.

Reply:
Yes, this could lead to undetermined states similar to unexpected error out which we discussed before in this issue #552. I will rethink about this and consider to delete it at per-component level for now.

Reply:
removed.